helix-core/src/main/java/org/apache/helix/common/caches/TaskDataCache.java

```diff
@@ -267,6 +277,10 @@ public void updateWorkflowContext(String resourceName, WorkflowContext workflowC
    * Update context of the Workflow or Job
    */
   private void updateContext(String resourceName, ZNRecord record) {
+    if (record != null && _initialContextMapHashcode.containsKey(resourceName)
+        && _initialContextMapHashcode.get(resourceName).equals((record.toString().hashCode()))) {
+      return;
+    }
```

Comment:
The problem I see here is that hash code equals does not mean the content is equal, right? 

hashcode1 ! = hashcode2 ==> content1 != content2: this is fine

hashcode1 == hashcode2  ==> content1 == content2: this is not true.  Unless you use security hash, the probability is very high that this is true.

Or is it that I did not fully understand the intention here?

Reply:
Yeah, your point is true since the possible input of hashcode is bigger than the possible outputs. However, since we convert the input to a string, the chance is pretty low (note that it will not be random strings and input string is derived from the ZNRecord). We had other options like deep copy and etc. but that options could cause memory and GC issues.

Reply:
Can we do something like if hash equals, let to do a further check of content equality before skipping it?  otherwise, this code has the potential of losing update, which would be hard to debug in the field.

Reply:
That is not possible or costly. We do in place change the content. If we want to check the content we need to do a deep copy of the context caches at the refresh phase which increases the memory footprint a lot.

Reply:
Then what is the solution here? 

Is it possible to calculate security hash using say md5?

Reply:
@dasahcc @jiajunwang  what do you think about this discussion? I think the concern is valid in theory but not in practice. http://sigpwned.com/2018/08/10/string-hashcode-is-plenty-unique/

Reply:
We can also use different algorithms such as md5. I feel it is kind of overkill but if there is a strong preference for that we can use different algorithms.

Reply:
I think MD5 is probably ok while hashcode is not. The thing is that even in you doc you quote, it has around 'An aggregate collision rate of 8 per 10,000 still isn’t bad in an absolute sense.'  0.1% is still too high for our usage. 

MD5 etc secure hash is the foundation of digital signature. The have much much lower false rate. Basically people can claim for all practical purpose, they can replace equality test. 

We can talk about it tomorrow.

Reply:
I added the MD5 commit now. Both of them work for me and IMO both would work considering our input is not a random string. Anyway, we can go with either of them. If we decided to go with MD5 we can keep the last commit. If not we can revert it.